### DISCUSSION TITLE
Loading data from various .hdf5 files in parallel

### DISCUSSION THREAD

**Discussion by joernwichmann** [firedrakeproject/firedrake#3371] [Firedrake support]
Hello everyone,

I am looking for help on the following topic. Let's say an oracle provides us with approximations to a stochastic PDE. It is stored in the following data format
![storage_structure](https://github.com/firedrakeproject/firedrake/assets/116003022/c9cac755-b16d-4031-8134-8586de575745)
Each level directory contains many .hdf5 files containing sample trajectories. 

My job is to load the approximations, so that further processing steps can be taken. I have implemented a sequential loader that works fine. However, it only uses one core, drastically limiting the loading speed. I aim to increase the loading speed by using multiprocessing. But unfortunately my first attempt fails. Let me guide you through this attempt and where it fails.

First I define space and time discretisation classes encapsulating discretisation variables. 
![disc_classes](https://github.com/firedrakeproject/firedrake/assets/116003022/05ce6b18-b55a-415d-8923-9fd4aee43842)

Next, a generic loader is defined by using the load function of _firedrake.CheckPointFile_.
![generic_loader](https://github.com/firedrakeproject/firedrake/assets/116003022/4f97221d-7162-45e0-b9e5-141280e7bc72)

The generic loader is used to define a specific loader using the storage directory structure.
![specific_loader](https://github.com/firedrakeproject/firedrake/assets/116003022/f0eed99c-5665-41fa-8c91-3f2fb07aca6a)

Now, we can either use the specific loader sequentially or in parallel.

The sequential load just iterates over the requested samples. 
![load_seq](https://github.com/firedrakeproject/firedrake/assets/116003022/78b9ea2e-394d-4ae6-bbe3-3b2ebee2e449)
The sequential loader works fine.

The parallel load uses the package _concurrent.futures_. 
![load_par](https://github.com/firedrakeproject/firedrake/assets/116003022/6a6beb29-9edb-494c-9f57-0e3a40583a2a)
Unfortunately, when I run the parallel load, I get the following error message.
![error](https://github.com/firedrakeproject/firedrake/assets/116003022/6e7bd587-0a13-43b3-8a8c-2ff2f6dd0b90)

Q: Do you have any idea how to fix the parallel load?

**Comment by connorjward** [comment]
Firedrake uses MPI to handle parallelism. To get parallel HDF5 loads of CheckpointFiles you simply need to to execute `mpiexec -n #procs python myfile.py` instead of `python myfile.py`. The sequential and parallel codes should be identical.

If you want the ranks to act independently you might want to use [ensemble parallelism](https://www.firedrakeproject.org/parallelism.html#ensemble-parallelism) instead.

Does this answer your question?

**Reply by joernwichmann** [reply]
Do I understand it correctly, ` mpiexec -n #procs python myfile.py` executes myfile.py n-times without any communication inbetween? This is not what I need. 

I want to have local loaders that load partial data. Afterwards the partial data should be sent to a master process that assembles the partial data to a global object.  Can I use _firedrake.Ensemble_ to communicate generic objects rather than _firedrake.Function_  s?

**Reply by connorjward** [reply]
`mpiexec` launches a number of processes that each run the same Python script. These processes *are* able to communicate with each other since they share an MPI communicator (`mpi4py.MPI.COMM_WORLD`). For a Firedrake simulation what this means is that each process only sees a local portion of the full mesh. This is done totally transparently to the user. You should just be able to wrap your serial script in an `mpiexec` and we will automagically distribute the mesh and handle all communication.

`CheckpointFile`s work in the same way. You can load the file on any number of processes and each process will then only see its partitioned section of the mesh.

I don't understand from your example whether you are trying to (a) load a single `CheckpointFile` shared between processes, or (b) load multiple `CheckpointFile`s, one per process. `mpiexec` will immediately give you (a), for (b) you would need ensemble parallelism.

**Reply by joernwichmann** [reply]
I am interested in (b') load multiple `CheckPointFile`s, many per process. I'll have a look at ensemble parallelism. Thanks.
